Related objects listing on OPTIONS/DELETE

neratneto opened the pull request:
Comes from the django admin deletion process.

neratneto committed:
Function to display related objects

neratneto committed:
Update related_objects.py

neratneto committed:
Fixing indentation errors from flake8

neratneto committed:
Added a Mixin

Creation of the RelatedObjectsModelMixin, which allows the user to use the related_objects function through a http OPTIONS request.

tiagovizoto commented:
@neratneto Não passou no flake8.

neratneto committed:
Fixing flake8 indentation problem

adolfosabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,39 @@
+        yield from collector.instances_with_model()
+
+
+class RelatedObjectsModelMixin(View, models.Model):
```
É uma `View`, ou é uma `Model`?

adolfosabino replied:
Creio que seria uma `DetailEndpoint`

adolfo-sabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,39 @@
+        class Meta:
+            abstract = True
+
+        allowed_methods = ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS']
```
Me parece estranho um `mixin` definir esses atributos

adolfosabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,39 @@
+
+            response = HttpResponse()
+            response['Allow'] = self.allowed_methods
+            response['Related-Objects'] = related
```
Se não me engano
```response['Related-Objects'] = related_objects(self.get_instance(request, *args, **kwargs))```
devia resolver

neratneto replied:
Ele retorna um `generator` quando chamado desse jeito. Mas realmente simplifica se fizer diretamente uma `list()` a partir dele.

adolfosabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,39 @@
+            related = []
+            for model, obj in related_objects(DetailEndpoint.get_instance(self, request, *args, **kwargs)):
+                related.append(obj)
+                related.append(model)
```
Estranho uma lista não homogênea. Geralmente agrupamos esses pares em tuplas

adolfosabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,39 @@
+                return Http400('Method Not Allowed')
+
+            related = []
+            for model, obj in related_objects(DetailEndpoint.get_instance(self, request, *args, **kwargs)):
```
De onde saiu esse DetailEndpoint? Sem instanciar ainda :thinking:

neratneto committed:
Simplifying code and fixing details

neratneto committed:
Update related_objects.py

adolfosabino reviewed (dismissed):
:+1:

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,32 @@
+        yield from collector.instances_with_model()
+
+
+class RelatedObjectsModelMixin(DetailEndpoint):
```
The name should not contain "Model", since it's all about an endpoint.

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,32 @@
+        To use it this mixin, make a http OPTIONS request to the location you want to analyze.
+        """
+
+        class Meta:
```
This class is for Django models, not for endpoints.

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,32 @@
+        class Meta:
+            abstract = True
+
+        allowed_methods = ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS']
```
I don't think a mixin should set these methods. The "final" class is the one that should add `OPTIONS` to its own `allowed_methods`.

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,32 @@
+
+            response = HttpResponse()
+            response['Allow'] = self.allowed_methods
+            response['Related-Objects'] = list(related_objects(self.get_instance(self, request, *args, **kwargs)))
```
It should be contained in the response **body**.

neratneto committed:
Improving code

Fixed name of mixin; unnecessary class Meta; removed allowed_methods; and made it a body response.

laylacodogno commented:
@neratneto ta em wip ainda? ou podemos revisar?

laylacodogno reviewed (approved):
👍

neratneto committed:
Addition of a test

adolfosabino commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,25 @@
+            if 'OPTIONS' not in self.allowed_methods:
+                return Http400('Method Not Allowed')
+
+            response = HttpResponse(list(related_objects(self.get_instance(self, request, *args, **kwargs))))
```
Tem algumas coisas erradas nesta linha:

1- É preferivel quebrar cadeias de chamadas de métodos/funções em mais *statements* para facilitar a compreensão, e adicionando mais "pontos de quebra", facilita a eventual depuração.
2- Essa função `related_objects` é chamada somente aqui. Não é melhor que ela já retorne o tipo a ser usado aqui? (no caso lista). A princípio não há motivos pra ser um generator.
3- Tá passando o `self` duas vezes. uma no `self.get...` e outra como primeiro argumento explícito. Isso passa mesmo no teste? o-o

Arrumando essas queestões a instanciação dessa `response` ficaria:

```
instance = self.get_instance(request, *args, **kwargs)
response = HttpResponse(related_objects(instance))
```
Mais legível?
Não tente economizar linhas!

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,28 @@
+
+
+def related_objects(target):
+        collector = Collector(None)
```
Indented with **8** spaces?

cleberzavadniak commented on powerlibs/django/contrib/related_objects.py:
```diff
@@ -0,0 +1,28 @@
+
+            instance = self.get_instance(request, *args, **kwargs)
+
+            response = HttpResponse(related_objects(instance))
```
I believe you should call `self.get_related_objects`, `related_objects` becoming a **method** of this class.
